File: 9-linked-lists2/4-cycle-detection/main.py
# Cycle Detection - Python Example

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def detect_cycle(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = head.next  # cycle

print("Cycle detected:", detect_cycle(head))

File: 9-linked-lists2/4-cycle-detection/main.cpp
// Cycle Detection - C++ Example

#include <iostream>

struct Node {
    int data;
    Node* next;
    Node(int val) : data(val), next(nullptr) {}
};

bool detectCycle(Node* head) {
    Node* slow = head;
    Node* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return true;
    }
    return false;
}

int main() {
    Node* head = new Node(1);
    head->next = new Node(2);
    head->next->next = new Node(3);
    head->next->next->next = head->next;  // cycle

    std::cout << "Cycle detected: " << (detectCycle(head) ? "true" : "false") << std::endl;

    return 0;
}
